Add a SBOM file in CycloneDX format

hughsie opened the pull request:
Hi,

My name is Richard Hughes and I'm a developer at Red Hat. I'm the maintainer of fwupd and LVFS, and am trying to improve software supply chain security by encouraging OEMs, ODMs and IBVs to ship Software Bill of Materials with each firmware binary blob (SBOMs).

I'm working alongside lots of other companies proactively trying to do the right thing. The reason I've opened this pull request is because your project is either used in the *build process* of a firmware we care about (e.g. EDK II, or coreboot) or is built *into the firmware binary* itself. Although my personal focus is on firmware, the SBOM file is in CycloneDX format (one of the most popular industry standards) which makes it also useful when building containers or OS images too.

I would like to contribute this template SBOM file into your project that gets included into source control with substituted values that get populated automatically. I'm not super familiar with your project, and so I've done my best populating the project values -- but please point out any that are incorrect and I'll fix them up. I've also put the `sbom.cdx.json` file in what I feel is the right place, but please say if you want me to put it somewhere different or name it a different thing; the directory and `sbom` prefix are unimportant.

I've written a bit more about this proposal here https://blogs.gnome.org/hughsie/2024/11/14/firmware-sboms-for-open-source-projects/ and there's also lot more information about firmware SBOMs here: https://lvfs.readthedocs.io/en/latest/sbom.html – many thanks for your time.

hughsie committed:
Add a SBOM file in CycloneDX format

Improve supply chain security by including a SBOM file with substituted values.

This will be used to construct a composite platform SBOM.

Signed-off-by: Richard Hughes <richard@hughsie.com>

levitte commented:
Hmmmm. Aren't SBOMs related to release packages? And isn't there tooling to generate them?

I'm a little surprised to see this relatively raw form, where I'm sure there must be tooling to generate a completely SBOM as part of building/installing/packaging.

hughsie commented:
Yes, the SBOM is generated as part of the firmware build. What this file is doing is basically making the guesses better -- i.e. we know we're building libprov, but we don't know the license, the homepage or what to call the team providing the library. 

The uswid tools takes these incomplete SBOM files and then builds them into a complete composite SBOM to represent the firmware. Also, having an upstream reference to what the PURL and CPE values should be means we have something we can trust; I could quite easily spin up a web-service that we say "_what CPE do we use for libprov_" -> "_cpe:2.3:a:provider-corner:libprov:\*:\*:\*:\*:\*:\*:\*:\*_` but we don't actually know if that's still true, up to date, or what the maintainer actually wants them to be. Putting the template upstream means we can *trust* the values we find in the checked out code during the build process.

levitte commented:
Hmm OK. Thing is, this raises a concern.

In this example, I'm referred to as a supplier. But am I? I think not, 'cause I am not making any package, and I don't know what fork is used, or even if the final producer has made any in-house changes.

This means I cannot accept this in its current form. It seems to me that the bom-ref, cpe and externalReferences values may need to be generated (not sure by who) depending on what fork is used, and the supplier info by whoever decides to take on that role (packager, or in your scenario, posdibly the firmware producer itself).

hughsie commented:
> In this example, I'm referred to as a supplier. But am I?

I think you are; in some projects they've asked us to use things like `The Mbed TLS Contributors` which is 100% fine if you'd prefer that.

> I don't know what fork is used, or even if the final producer has made any in-house changes.

In this case the expectation is that the SBOM file would be rewritten -- e.g. using https://github.com/hughsie/python-uswid?tab=readme-ov-file#ini-file -- it would be disingenuous at best to say you're the author when someone's done something to your project you disapprove of. In the case of libprov the dep chain is `libprov` -> `gost-engine` -> `OpenSSL` -> `EDK2` -> `OEM firmware` -- and in this common case it appears to be exactly your code with no downstream changes.

levitte commented:
> > In this example, I'm referred to as a supplier. But am I?
> 
> I think you are

I guess we disagree. However, this was a bit enlightening:

> it would be disingenuous at best to say you're the author when someone's done something to your project... 

Here, you talk about authorship, which isn't the same as supplier, also made quite visible in the file of this PR. I have nothing at all against being mentioned as the or along the authors.

hughsie commented:
> Here, you talk about authorship, which isn't the same as supplier

This is where there's a slight impedance mismatch between how open source developers see themselves (myself included btw) and how CISA and the other organisations supporting SBOMs define authors. In the SPDX, SWID and CycloneDX SBOM specifications the author in the SBOM is the person that _wrote the SBOM file_ itself, and the supplier is what we would think of as the "_author or maintainer of the upstream software_". I'd be 100% okay if you want to be marked as the author of the SBOM file _as well_, but from a SBOM PoV you're the supplier.

levitte commented on sbom.cdx.json:
```diff
@@ -0,0 +1,39 @@
+      "name": "libprov",
+      "version": "@VCS_VERSION@",
+      "description": "Small library of helpers for OpenSSL 3 providers",
+      "authors": [
+        {
+          "name": "@VCS_SBOM_AUTHORS@"
+        }
+      ],
```
If I understand your explanation of `@VCS_SBOM_AUTHORS@` correctly, this should be `metadata.authors`, not `components.[].authors`, and `components.[].authors` should be the set of component authors, not the SBOM authors.  It would be nice if there was a `@VCS_AUTHORS@` for which the result of `git shortlog -n -s` was used.

Refs: 
https://cyclonedx.org/docs/1.6/json/#metadata_authors
https://cyclonedx.org/docs/1.6/json/#components_items_authors

hughsie replied:
I think you're dead right about the `metadata.authors` change. I've got 5 pull requests to modify! :)

>  It would be nice if there was a @VCS_AUTHORS@

Great idea: https://github.com/hughsie/python-uswid/pull/66

levitte commented on sbom.cdx.json:
```diff
@@ -0,0 +1,39 @@
+          "name": "@VCS_SBOM_AUTHORS@"
+        }
+      ],
+      "supplier": {
+        "name": "Richard Levitte",
+        "url": [
+          "https://github.com/levitte"
+        ]
+      },
```
Please remove this.  It's not mandatory, and it's not necessary.

hughsie replied:
Done.

levitte commented:
> ... and the supplier is what we would think of as the "_author or maintainer of the upstream software_".

It's interesting, then, that the CycloneDX spec says this on the subject:

> The organization that supplied the component. The supplier may often be the manufacturer, but may also be a distributor or repackager.

Ref: https://cyclonedx.org/docs/1.6/json/#components_items_supplier

hughsie commented:
> It's interesting, then, that the CycloneDX spec says this on the subject:
>    The organization that supplied the component. The supplier may often be the manufacturer, but may also be a distributor or repackager.

I think from 40,000ft view you are the legal _supplier_, but I do understand why it makes you feel uncomfortable -- **it does me too**. I've made `authors` an alias for importing, and switched around the libprov one to use that instead.

levitte commented:
This PR now looks acceptable, but if you don't mind, place `sbom.cdx.json` in a `contrib/` subdirectory.

I would strongly suggest using this same resulting template for other projects you approach this way.

levitte committed:
Move sbom.cdx.json to contrib/

levitte committed:
contrib:sbom.cdx.json: change @VCS_TAG@ to @VCS_COMMIT@

Since libprov isn't really packaged, and isn't tagged, it's better
to refer to specific commits.

levitte commented:
I've added a few final touches (including moving to contrib/).  If you agree, this can be merged

hughsie commented:
Yup, looks good to me, thanks! I've added a section about author/supplier in the template I'm sending to projects. Thanks for all your help, it's much appreciated.

levitte commented:
Thank you for listening
